### ISSUE TITLE
Tokio spawn issue for websocket impl

### ISSUE THREAD

**Issue by CedricCouton** [Dushistov/couchbase-lite-rust#94]
Hi,

Regularly in our production we have some log that indicate that the socket is closed when the ws_write try to write in the socket. After analyse the log of the sync gateway we have two cases : invalid checksum and error during decompression. I focused on the first one to start. Invalid checksum is throw by the blip protocol. Checksum is cumulative so the value depends of all the precedent messages. When I deep dive in the log I can see that there is sometimes missing messages (MSG#) : bump from msg#6 to msg#7 or 8 for example. When I tried to reproduce I didn't success to.  It's a very edge case. After some loop + logs I finally reproduced it and figured out the reason. In the ws_write function the Tokyo::spawn is used but it don't guarantee the order of the execution of the future spawned. so sometimes we can have the bad order of messages. I think we can use here a block_on wdyt? I will submit a pr for that soon. Cheers

**Comment by Dushistov** [comment]
>  I think we can use here a block_on wdyt?

`block_on` has many downsides, it panics if called during runtime shutdown,
and has real issues with single thread runtime.

**Comment by CedricCouton** [comment]
Yes no doubt about that, it's just a proposition. How can we solve the issue :-) ?

**Comment by CedricCouton** [comment]
Maybe we can have a queue adding the message inside it and pop a message in each spawned future?

**Comment by Dushistov** [comment]
> Maybe we can have a queue adding the message inside it and pop a message in each spawned future?

Or we can spawn one task during socket open that read channel and write to socket in the loop.
May be reuse `main_read_loop` for that. There is only one thing that tricky is properly handle closing.
Because of we need send close ack.
